test/OpenTelemetry.Instrumentation.SqlClient.Tests/SqlClientTests.cs

```diff
@@ -56,6 +56,27 @@ public void Dispose()
             GC.SuppressFinalize(this);
         }
 
+        [Fact]
+        public void AddSqlClientInstrumentationNamedOptionsSupported()
+        {
+            int defaultExporterOptionsConfigureOptionsInvocations = 0;
+            int namedExporterOptionsConfigureOptionsInvocations = 0;
+
+            using var tracerProvider = Sdk.CreateTracerProviderBuilder()
+                .ConfigureServices(services =>
+                {
+                    services.Configure<SqlClientInstrumentationOptions>(o => defaultExporterOptionsConfigureOptionsInvocations++);
+
+                    services.Configure<SqlClientInstrumentationOptions>("Instrumentation2", o => namedExporterOptionsConfigureOptionsInvocations++);
+                })
+                .AddSqlClientInstrumentation()
+                .AddSqlClientInstrumentation("Instrumentation2", configureSqlClientInstrumentationOptions: null)
```

Comment:
I think the goofy thing about adding instrumentation twice is that there are a couple strange side effects that could occur.

With the SqlClient instrumentation I _think_ what would happen is that multiple activities would be started

https://github.com/open-telemetry/opentelemetry-dotnet/blob/928d77056c1d353d8ba72ad3b4b565398117b352/src/OpenTelemetry.Instrumentation.SqlClient/Implementation/SqlClientDiagnosticListener.cs#L59-L64

If this pattern were applied to the ASP.NET Core instrumentation I think the side effect might be different. I _think_ there'd only be one activity (the one created by ASP.NET Core), but the instrumentation would be invoked twice... maybe duplicating work that shouldn't be duplicated?

Reply:
Good points! Just to clarify, it was always possible before to register instrumentation(s) multiple times. All I'm doing here is supporting binding named options to it. We should look at this as a separate effort?

Reply:
Yea, definitely a separate effort to consider how and if we should prevent multiple calls to add instrumentation.

Though, if we did prevent multiple calls then would there value in supporting named options for instrumentation?

Reply:
I don't have a well thought out answer 😄 Really all I'm doing on these PRs is establishing anytime you ask for options, pass a name (either user-supplied or default). IMO it is a good pattern that allows the users to have control if they happen to need it without a lot of effort to do it.

Why would they need it? Off the top of my head...

* Using named options always could just be a preference thing for large complex applications?

* Imagine some application with 2 tracer providers. That application has a single `IConfiguration` which needs to apply to different providers. Naming helps in this case in a big way. More clear with exporters for sure. Instrumentation... depends on what the behavior is for calling "AddInstrumentation" across two providers. What the user _probably_ wants to do in that case is create one activity, listen on both providers? But then we also have filter/enrich things to worry about. Kind of undefined at the moment.

* Reaching here, but thinking about the OTel config file thing. Let's say we read in OTel standard config file and expose via Options/IConfiguration keys so that SDK can pick it up it. Having the name available everywhere might be useful as a disambiguation/collision avoidance mechanism?